src/Datasource/QueryInterface.php

```diff
@@ -0,0 +1,362 @@
+     *
+     * @param string $field The field to alias
+     * @param string $alias the alias used to prefix the field
+     * @return array
+     */
+    public function aliasField($field, $alias = null);
+
+    /**
+     * Fetch the results for this query.
+     *
+     * Will return either the results set through setResult(), or execute this query
+     * and return the ResultSetDecorator object ready for streaming of results.
+     *
+     * ResultSetDecorator is a traversable object that implements the methods found
+     * on Cake\Collection\Collection.
+     *
+     * @return \Cake\Datasource\ResultSetInterface
+     */
+    public function all();
+
+    /**
+     * Populates or adds parts to current query clauses using an array.
+     * This is handy for passing all query clauses at once. The option array accepts:
+     *
+     * - fields: Maps to the select method
+     * - conditions: Maps to the where method
+     * - limit: Maps to the limit method
+     * - order: Maps to the order method
+     * - offset: Maps to the offset method
+     * - group: Maps to the group method
+     * - having: Maps to the having method
+     * - contain: Maps to the contain options for eager loading
+     * - join: Maps to the join method
+     * - page: Maps to the page method
+     *
+     * ### Example:
+     *
+     * ```
+     * $query->applyOptions([
+     *   'fields' => ['id', 'name'],
+     *   'conditions' => [
+     *     'created >=' => '2013-01-01'
+     *   ],
+     *   'limit' => 10
+     * ]);
+     * ```
+     *
+     * Is equivalent to:
+     *
+     * ```
+     *  $query
+     *  ->select(['id', 'name'])
+     *  ->where(['created >=' => '2013-01-01'])
+     *  ->limit(10)
+     * ```
+     *
+     * @param array $options list of query clauses to apply new parts to.
+     * @return $this
+     */
+    public function applyOptions(array $options);
+
+    /**
+     * Compiles the SQL representation of this query and executes it using the
+     * configured connection object. Returns the resulting statement object.
+     *
+     * Executing a query internally executes several steps, the first one is
+     * letting the connection transform this object to fit its particular dialect,
+     * this might result in generating a different Query object that will be the one
+     * to actually be executed. Immediately after, literal values are passed to the
+     * connection so they are bound to the query in a safe way. Finally, the resulting
+     * statement is decorated with custom objects to execute callbacks for each row
+     * retrieved if necessary.
+     *
+     * Resulting statement is traversable, so it can be used in any loop as you would
+     * with an array.
+     *
+     * This method can be overridden in query subclasses to decorate behavior
+     * around query execution.
+     *
+     * @return \Cake\Database\StatementInterface
+     */
+    public function execute();
```

Comment:
This returns a ```StatementInterface``` would it make sense to return something else here? ```StatementInterface``` has to do with databases.